Expose After Render Callback from directives

hoebbelsB opened the pull request:
# WIP

Todod:
- [x] Implement renedered$
- [x] Implement scheduleCD
- [x] Implement rxScheduleCD
- [x] Demonstrate the different ways to access the signal
- [ ] Demonstrate/Document Edge cases and cavets

# Description

Implemented a new `Output` for the `LetDirective` which emits the latest value after it was rendered by the strategy. The feature was requested as _afterRender signal_ for the upcoming `@rx-angular/template` RC release #203.

# Why

The signal emits after the `RenderStrategy` has detected the changes to be rendered. Users now have a proper notification about when a change was actually reflected to the view. This is especially useful for DOM calculations. E.g. where you rely on the height of an element after it's state got manipulated. 

I have implemented a small example use case demonstrating this behavior. In this example I am calculating the height of a div after it's state chanted. The height gets calculated twice. One time after the value got chanted and one time after the renderCallback emitted. 


![image](https://user-images.githubusercontent.com/4904455/93690841-added600-fadd-11ea-91e2-b73bc8ca65fa.png)


Please test it out:

* yarn nx s examples
* navigate to `render-callback/render-callback-01`
* press the "update content" button


# Usage

## Template Syntax

Since it's not possible to subscribe to output-bindings of structural directives with `*-Syntax`, users would have to use `TemplateSyntax` to bind to the callback inside the view.

```html
    <ng-template let-content
                 [rxLet]="content$"
                 (rendered)="rendered$.next($event)">
      <div class="example-box"
           #box>
        {{ content }}
      </div>
    </ng-template>
```

## ViewChild

Another way would be to fetch the `LetDirective` as `ViewChild` and subscribe to the event manually.

```ts
  @ViewChild(LetDirective) letDirective: LetDirective<any>;
  readonly rendered$ = this.letDirective.rendered;
```

# Known Issues & Caveats

The implementation relies on the `RenderStrategy` to work properly. So we have to make sure that all of our strategies are emitting after they got rendered.

How is it possible for the `PushPipe` to have an output? (inside the view. you could always inject it as viewchild)

Another thing to think about is `coalescing`. As long as the user only has 1 `LetDirective` which handles changeDetection for the component, there shouldn't be an issue. But as soon as multiple `renderAware` or even `RenderStrategy` users are in place, things can quickly get messy.
There can be a situation where you subscribe to the renderSignal of 1 particular LetDirective, which would never emit a value, because those changes are detected by another LetDirective/RenderStrategy/PushPipe.

To encounter this issue we could explain that you would have to listen to all Let/Push/RenderStrategy in order to get notified that your component re-rendered.

But I guess this won't be an issue anymore after #294 was transformed into real code :+1: 

I am happy for any feedback and discussions on this topic!

BioPhoton commented:
Hey @hoebbelsB!

I really appreciate your nice and detailed description. The comments regarding multiple changes in one component especially! 👍

I can imagine after the switch to EmbeddedView#detectChanges every directive will fire a value as we have a scope per EView.
WDYT?

BioPhoton commented on libs/template/src/lib/core/render-aware/render-aware_creator.ts:
```diff
@@ -94,7 +95,8 @@ export function createRenderAware<U>(cfg: {
         )
-    )
+    ),
+    publishReplay(1)
```
We have to use publish here as we would create confusing output if new subscribers are there.
```suggestion
    publish()
```

BioPhoton commented on libs/template/src/lib/core/render-aware/render-aware_creator.ts:
```diff
@@ -5,12 +5,12 @@ import {
   distinctUntilChanged,
   filter,
-  map,
+  map, publishReplay,
```
See comment below
```suggestion
  map, publish,
```

BioPhoton commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -306,6 +307,8 @@ export class LetDirective<U> implements OnInit, OnDestroy {
     this.templateManager.addTemplateRef('rxSuspense', templateRef);
   }
 
+  @Output() readonly rendered = defer(() => this.renderAware.rendered$);
```
```suggestion
  // We use defer here as the as otherwise the the `@Output` decorator subscribes earlier than the arnderAware property is assigned
  @Output() readonly rendered = defer(() => this.renderAware.rendered$);
```

BioPhoton reviewed (commented):
I suggested some tiny changes regarding documentation.
I'm super hyped about this PR!

What we can do is add:
- API docs
- Usage examples in the real docs

Not sure if we should put it in the same PR... As u wish.

BioPhoton commented:
Regarding global strategy and render callback:

We should throw an error ATM until we found a way to detect this.

BioPhoton commented:
I spent some time on the output of renderEffect$. 
I implemented materialize to have feedback for all RxJS notifications as well as added a suspense notification (this should be challenged). 
I also created a `RxNotification` type for the return value.

I started to adopt strategies to follow this principle.
💡 `global` needed `animationFrameTick` and `native` needed `setTimeoutTick`. 
💡 For `noop` I ignored all values forwarded.

I also thought about the `scheduleCD` method which also needs to follow this change. I guess we don't test this case ATM.

hoebbelsB commented:
> Regarding global strategy and render callback:
> 
> We should throw an error ATM until we found a way to detect this.

already resolved :)

hoebbelsB commented:
> I spent some time on the output of renderEffect$.
> I implemented materialize to have feedback for all RxJS notifications as well as added a suspense notification (this should be challenged).
> I also created a `RxNotification` type for the return value.

Thx, I have extended it being generic + removed not needed fields of the original `Notification`.

> I started to adopt strategies to follow this principle.
> bulb `global` needed `animationFrameTick` and `native` needed `setTimeoutTick`.
> bulb For `noop` I ignored all values forwarded.

THX!

> I also thought about the `scheduleCD` method which also needs to follow this change. I guess we don't test this case ATM.

I have implemented all of the static methods + wrote tests

BioPhoton commented:
I thought about the firehouse tests. I guess it works because we had this issue some time ago, but do we have tests for that situation?

BioPhoton commented:
Isn't it possible to use exportAs in the directives decorator to expose the signal?

kajetansw commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -303,14 +294,25 @@ export class LetDirective<U> implements OnInit, OnDestroy {
+  // is assigned
+  @Output() readonly rendered = defer(() => this.renderAware.rendered$.pipe(
+    filter(({ kind }) => this.templateManager.hasTemplateRef(kind)),
+    map(({ value }) => value as U),
```
If I'm not mistaken the `value` is already inferred as `U` (at least my IDE says so). Please check it out, maybe this `as` clause won't be necessary ;)

hoebbelsB replied:
ty, looks good

kajetansw commented on libs/template/src/lib/core/render-aware/interfaces.ts:
```diff
@@ -1,5 +1,9 @@
+
+export type RxNotificationKind = 'S' | 'N' | 'E' | 'C';
+type NotificationExtract = 'value' | 'hasValue' | 'error';
+export type RxNotification<T> = Pick<Notification<T>, NotificationExtract> & { kind: RxNotificationKind };
```
Nice! I didn't know RxJS provides its own notification markings!

kajetansw commented on libs/template/src/lib/core/render-aware/render-aware_creator.ts:
```diff
@@ -104,40 +108,38 @@ export function createRenderAware<U>(cfg: {
+
 function renderWithLatestStrategy<T>(
   strategyChanges$: Observable<RenderStrategy>
 ): MonoTypeOperatorFunction<T> {
```
```suggestion
): OperatorFunction<T, RxNotification<T>> {
```

Seems like the return type change a bit with your changes.

hoebbelsB replied:
thx! i will merge it manually, since your suggestion is missing the "import" statement for OperatorFunction

BioPhoton reviewed (changes requested):
Hi as a last missing piece we need to document clearly when the edge case with renderCD and coalescing occurs.

Let me know if u need help!

hoebbelsB commented on libs/template/src/lib/render-strategies/util/util.ts:
```diff
@@ -0,0 +1,29 @@
+    afterCD();
+  }
+}
+
```
done

hoebbelsB commented on libs/template/spec/render-strategies/render-strategies/local.strategy.spec.ts:
```diff
@@ -54,21 +56,23 @@ describe('local Strategy', () => {
     });
 
-    it(`should call cdRef#detectChanges 1 times when scheduleCD is called multiple times sync`, (done) => {
+    it(`should call cdRef#detectChanges 1 times and afterCD 1 time when scheduleCD is called multiple times sync`, (done) => {
```
what are firehouse tests?

BioPhoton replied:
I learned it from @cartant 

>It tests whether or not a subscribers to synchronous sources - like range without a scheduler - stop listening to said source upon unsubscription within the next handler - e.g. a range of 10 elements should stop generating after 3 elements if a take(3) operator is applied to it.

cartant replied:
_Firehose_, not firehouse.

![firehose](https://user-images.githubusercontent.com/3878593/95398779-09b4b600-094a-11eb-89b7-428bb62495e0.gif)

hoebbelsB replied:
haha :+1:

hoebbelsB commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -395,10 +399,16 @@ export class LetDirective<U> implements OnInit, OnDestroy {
   }
 
+  private subscribeRenderCallbacks(): void {
+    this.subscription.add(this.rendered.pipe(
```
good point. implemented that :+1:

hoebbelsB commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -303,14 +294,26 @@ export class LetDirective<U> implements OnInit, OnDestroy {
     templateRef: TemplateRef<LetViewContext<U | undefined | null> | null>
   ) {
-    this.templateManager.addTemplateRef('rxSuspense', templateRef);
+    this.templateManager.addTemplateRef('S', templateRef);
```
so we don't want to be consistent with rxjs Notification? I quite like the 'S' approach. Or is there a downside I don't get?

BioPhoton replied:
At the moment we have the following naming:

```typescript
// RenderAware
type RxNotificationKind = 'S' | 'N' | 'E' | 'C';

// Let Directive - template names
type RxTemplateName = 'rxNext' | 'rxComplete' | 'rxError' | 'rxSuspense';

// Let Directive - value observer
interface RxTemplateObserver<T> {
  suspense?: () => void;
  next?: (v: T) => void;
  error?: (v: any) => void;
  complete?: () => void;
}

// Let Directive - view context
interface RxViewContext<T> {
  $suspense: boolean;
  $implicit: T; // next
  $error: false | Error;
  $complete: boolean;
}
```

2 things I can say here:
- We could harmonize the naming a bit more across the codebase
- I think `rxNext` is more readable than 'N'

What do you say? :)

hoebbelsB replied:
yes, as for the public API, `rxNext` is better :+1: I will work on the refactoring

hoebbelsB replied:
So if we stick to this we get the following result:

```ts
// RenderAware
type RxNotificationKind = 'rxSuspense' | 'rxNext' | 'rxError' | 'rxComplete';

// Let Directive - value observer -> no more derivation of Observer possible
interface RxTemplateObserver<T> {
  rxSuspense?: () => void;
  rxNext?: (v: T) => void;
  rxError?: (v: any) => void;
  rxComplete?: () => void;
}

// Let Directive - template names
type RxTemplateName = 'rxSuspense' | 'rxNext'  | 'rxError' | 'rxComplete';

// TemplateManager - view context
interface RxViewContext<T> {
  $rxSuspense: boolean;
  $implicit: T; // next
  $rxError: false | Error;
  $rxComplete: boolean;
}

@Input()
  set rxLetRxComplete(
    templateRef: TemplateRef<LetViewContext<U | undefined | null> | null>
  ) {
    this.templateManager.addTemplateRef('rxComplete', templateRef);
  }

```

```html
<ng-container *rxLet="hero$; let hero; rxComplete: completeTemplate">
   <app-hero [hero]="hero"></app-hero>
<ng-container>
<ng-template #completeTemplate>
  <mat-icon>thumb_up</mat-icon>
</ng-template>
```

is this really what we want?

hoebbelsB replied:
@BioPhoton I have refactored everything according to plan, fixed tests, updated documentation and created some utils ready for re-usage. please have a look :)

hoebbelsB commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -303,14 +294,26 @@ export class LetDirective<U> implements OnInit, OnDestroy {
+  ));
+
+  private subscription = new Subscription();
+  private readonly templateManager: TemplateManager<LetViewContext<U | undefined | null>,
```
i really like the idea. this would be a breaking change though

BioPhoton replied:
We are beta! If we want to break stuff we should do it now :)

hoebbelsB commented:
@BioPhoton @kajetansw I have updated the docs (jsDocs inside of the letDirective). Could you please check them and provide some feedback?
Looking forward to w/e firehouse tests are and refactoring the ViewContext :D

kajetansw commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -77,7 +66,8 @@ export interface LetViewContext<T> extends RxViewContext<T>  {
  *   `ɵdetectChanges`)
  * - distinct same values in a row (`distinctUntilChanged` operator),
- * - display custom templates for different observable notifications (suspense, next, error, complete)
+ * - display custom templates for different observable notifications (rxSuspense, next, error, complete)
```
```suggestion
 * - display custom templates for different observable notifications (rxSuspense, rxNext, rxError, rxComplete)
```
Wouldn't this be more consistent of a description?

kajetansw commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -77,7 +66,8 @@ export interface LetViewContext<T> extends RxViewContext<T>  {
  * - distinct same values in a row (`distinctUntilChanged` operator),
- * - display custom templates for different observable notifications (suspense, next, error, complete)
+ * - display custom templates for different observable notifications (rxSuspense, next, error, complete)
+ * - notify about after changes got rendered to the template (Rendercallback)
```
```suggestion
 * - notify after changes were rendered to the template (RenderCallback)
```
Just a suggestion of a sentence that is possibly easier to read.

kajetansw commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -142,6 +132,64 @@ export interface LetViewContext<T> extends RxViewContext<T>  {
+ * You can choose between using the [Template syntax](https://angular.io/guide/template-syntax), injecting the
+ * `LetDirective` as `@ViewChild()` and subscribe the event manually or providing a RenderCallback on your own.
+ *
+ * Please note that due to the built in
```
```suggestion
 * Please note that due to the built-in
```

kajetansw commented:
It would be nice of you to also generate Markdown docs for the `LetDirective` and update the `let-directive.md` :)

BioPhoton reviewed (changes requested):
Nice refactoring!! Just some tiny nitty gritty details and questions! :)

hoebbelsB commented on libs/template/src/lib/core/utils/rx-materialize.ts:
```diff
@@ -0,0 +1,31 @@
+        console.error(error);
+      }
+    }),
+    map(({ value, hasValue, error, kind }) => ({
+      value,
+      hasValue,
+      error,
+      kind: notificationKindToRxNotificationKind(kind)
+    })),
```
you mean re-setting the original "kind" property and forward the result?

hoebbelsB commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -41,9 +30,9 @@ export interface LetViewContext<T> extends RxViewContext<T>  {
- * So if the incoming `Observable` emits its value lazily (e.g. data coming from `Http`), your template will be
- * rendered lazily as well. This can very positively impact the initial render performance of your application.
+ * of your component. The `LetDirective` will render its template and manage change detection after it got an initial
+ *   value. So if the incoming `Observable` emits its value lazily (e.g. data coming from `Http`), your template will
```
needs to be re-edited, you are right. I mean the value which comes out of the bound `Observable`

hoebbelsB commented on libs/template/src/lib/core/render-aware/interfaces.ts:
```diff
@@ -36,5 +36,6 @@ export interface RenderStrategy {
    * Calls the render method directly
    */
+  // @TODO This is needed in situations like
   detectChanges: () => void;
```
I can remember the discussion, didn't we remove it in another PR? If not, should it be part of this PR?

BioPhoton replied:
We can remove it here imho...

BioPhoton commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -41,9 +30,9 @@ export interface LetViewContext<T> extends RxViewContext<T>  {
- * of your component. The `LetDirective` will render its template and manage change detection after it got an initial value.
- * So if the incoming `Observable` emits its value lazily (e.g. data coming from `Http`), your template will be
- * rendered lazily as well. This can very positively impact the initial render performance of your application.
+ * of your component. The `LetDirective` will render its template and manage change detection after it got an initial
```
```suggestion
 * of your component or EmbeddedView. The `LetDirective` will render its template and manage change detection after it got an initial
```

BioPhoton commented on libs/template/src/lib/let/let.directive.ts:
```diff
@@ -142,6 +132,64 @@ export interface LetViewContext<T> extends RxViewContext<T>  {
+ * Please note that due to the built-in
+ *   [coalescing][https://github.com/rx-angular/rx-angular/blob/master/libs/template/docs/concepts.md] can cause this
+ *   callback different in situations where multiple `LetDirectives` are used to render the same
+ * `Component`. Make sure to subscribe to every instance in your component to make sure you don't miss render
+ * notifications.
```
```suggestion
 * `Component`. Make sure to subscribe to every instance in your component to avoid missing render
 * notifications.
```

BioPhoton commented:
Change request:

Regarding naming:

- It's quite a refactoring
- `RxTemplateObserver` is missing
- refactorings in methods rxMaterialize

Demo:
- folder and demo example 
- lines commented out that are not implemented 

All this should go into separate PRs to reduce the size and make it easier to reason about changes.

hoebbelsB commented:
@BioPhoton as discussed, I have moved everything regarding the renderCallback into pocs living inside the demo application. The refactoring, the changes made to the renderAware and the RenderStrategies are still in place, since they are needed anyway and don't touch the public API.
I have removed the docs aswell and saved them in a separate md file next to the demo use cases

BioPhoton commented:
Amazing!

Thanks a lot for the effort! This is a quite tricky thing and the scope was also big.So again thanks for the effort and time you spent on this.
